Biosys 85 species separate column

serge-gaia opened the pull request:
see https://decbugs.com/view.php?id=6674

serge-gaia committed:
WIP: BIOSYS-85 Load species name as separate fields

serge-gaia committed:
WIP: SpeciesNameParser skeleton (not implemented) + test_schema_species_name_parser skeleton (not implemented)

serge-gaia committed:
WIP: SpeciesNameParser implemented + happy path tests.
TODO: SpeciesObservation tests fail and The SpeciesObservationValidator should use the parser

serge-gaia committed:
WIP: All legacy tests passed (after some refactoring)
TODO: validators for the multicolumn species.

serge-gaia committed:
More tests

serge-gaia commented on biosys/apps/main/utils_data_package.py:
```diff
@@ -1081,6 +997,226 @@ def _find_site_code_field(self):
 
 
+class SpeciesNameParser(object):
+    """
```
The meat of this PR

tony-gaia commented on biosys/apps/main/fixtures/test-datasets.json:
```diff
@@ -52,14 +52,20 @@
                   "description": "",
                   "name": "Genus",
-                  "title": ""
+                  "title": "",
```
I thought these test fixtures were on their way out?

serge-gaia replied:
They are but slowly. Still used by some tests. I've refactored the species_observations tests to not use them.

tony-gaia commented on biosys/apps/main/utils_data_package.py:
```diff
@@ -1081,6 +997,226 @@ def _find_site_code_field(self):
+            value = self._cast_field(field, record)
+            if value:
+                result += ' {}'.format(value)
+        return result.strip()
```
Could alternatively use string ```join()``` function here rather than appending strings then stripping.

serge-gaia replied:
Yes the `join`. I forgot. I was looking for a `reduce` function

tony-gaia commented on biosys/apps/main/tests/api/test_species_observation.py:
```diff
@@ -2,95 +2,94 @@
+        ds = self.ds_1
+        client = self.custodian_1_client
+        data = {
+            'Species Name': 'Chubby Bat',
```
A classic!

tony-gaia commented on biosys/apps/main/tests/api/test_species_observation.py:
```diff
@@ -272,137 +269,143 @@ def test_options(self):
+        ds.record_queryset.delete()
         self.assertEqual(
-            client.post(url, data, format='json').status_code,
+            client.post(url, data=payload, format='json').status_code,
```
not really necessary to use keyword arg here

serge-gaia replied:
yep

tony-gaia commented on biosys/apps/main/tests/api/test_species_observation.py:
```diff
@@ -572,337 +574,364 @@ def test_species_name(self):
         """
-        Test that the species name is extracted from the data and saved in DB
-        :return:
+        Test that name extraction after a PUT method
```
PUT or PATCH? also you're using a post method below.

serge-gaia replied:
The post is just to create the initial record

tony-gaia commented on biosys/apps/main/tests/api/test_species_observation.py:
```diff
@@ -1462,3 +1533,154 @@ def test_species_name_collected_api_update(self):
+            {
+                "name": "Genus",
+                "type": "string",
+                "constraints": helpers.REQUIRED_CONSTRAINTS
```
Should it be ```REQUIRED_CONSTRAINT``` with no ```S```

tony-gaia commented on biosys/apps/main/tests/api/test_species_observation.py:
```diff
@@ -1462,3 +1533,154 @@ def test_species_name_collected_api_update(self):
+    species_facade_class = helpers.LightSpeciesFacade
+
+    @staticmethod
+    def schema_with_4_columns_genus():
```
Defined by not tested - I'll leave it to you as to whether to add another test for it.

tony-gaia commented on biosys/apps/main/tests/test_schema_date_parser.py:
```diff
@@ -0,0 +1,390 @@
+        descriptor = self.descriptor
+        parser = ObservationDateParser(descriptor)
+        self.assertTrue(parser.is_valid())
+        self.assertTrue(len(parser.errors) == 0)
```
I like to use ```assertEqual``` for equality testing

serge-gaia committed:
Fixes from code review
Added validation rule about name_id being of type integer.
Skip the name_id lookup if nameid = -1
